modules/mauth-common/src/main/java/com/mdsol/mauth/MAuthRequest.java

```diff
@@ -95,11 +130,19 @@ public MAuthRequest(String authenticationHeaderValue, byte[] messagePayload, Str
     this.appUUID = appUUID;
     this.requestSignature = requestSignature;
     this.messagePayload = messagePayload;
+    this.bodyInputStream = bodyInputStream;
     this.httpMethod = httpMethod;
     this.requestTime = requestTime;
     this.resourcePath = resourcePath;
     this.queryParameters = queryParameters;
     this.mauthVersion = MAuthHeadersHelper.getMauthVersion(authenticationHeaderValue);
+
+    // we always use inputStream for Auth,
+    // wrap byte array in ByteArrayInputStream if bodyInputStream doesn't set
+    if (bodyInputStream == null) {
+      if (messagePayload == null) messagePayload = new byte[0];
```

Comment:
I think not using a superclass like I suggested before is fine, but we need to make a clear distinction between:

* full message payload is not available (only stream)
* user specified full message payload (whether it's empty or not)

The first point is useful because it means you can actually abort V1 fallback auth with a meaningful message.

So I think what we need to do:

* In this primary constructor, throw an illegal argument exception if both messagePayload and bodyInputStream is not null
* In the constructor which takes a `byte[]` body, call the primary constructor with messagePayload = `new byte[0]` if the user passed in null as the body bytes
* V1 fallback validation should check that messagePayload == null, and if so just abort validation with a log message `"V1 authentication fallback is not available because the full request body is not available in memory"`

Reply:
Updated as suggested in https://github.com/mdsol/mauth-jvm-clients/pull/132/commits/54300d833460aad371e127fc8d8eeea8d10f69d9